### DISCUSSION TITLE
Using alpine with dynamically loaded components....

### DISCUSSION THREAD

**Discussion by rickghome** [alpinejs/alpine#3656] [3. Feature ideas]
### A Recipe That Should REALLY Be In The Alpine Docs

(NOTE: This has been edited for greater clarity).

We can agree that Alpine is amazing.   The doc does a great job of demonstrating how to use it to place components on a server-rendered page returned as an HTTP request response.. I've realized that it can also be used to inject components onto your page dynamically, if you're careful.  

Presume a scenario common to SPA app's where you're RESTfully calling your server to retrieve dynamic content (e.g. via fetch, Ajax, Axios, etc.).  But what if you want to inject content that is part of a RESTful response, that happens to contain an Alpine component?  What is necessary to properly initialize that dynamically rendered component to work correctly?  

The Alpine docs show this as a common (simplified) example for making a component:

```
<div x-data="foo()"> ... stuff ... </div>
<script>
    function foo() {
        return {
            field1 : value1,
            field2: value2,
            init() {console.log("initialized foo");},
            doSomething(event) { ...}, 
            doSomethingElse() {}
       }
   }
</script>
```

This pattern makes Alpine elegant and simple. But, it won't work if you attempt to return that fragment dynamically from your server in response to a REST call.  For security reasons, the <script> won't be executed, and Alpine won't initialize your component.

To get your component to work in the dynamically loaded scenario, the basic work-around is to pre-load the <script/> onto the page where you want to dynamically load a component. Presuming your server RESTfully returned an HTML partial, your client-side code can inject the partial onto the page. Since the script is already present, the x-data attribute will be picked up by Alpine, and your dynamically loaded component will be property initialized.

**Comment by ekwoka** [comment]
Or, just inline the data itself with a data agnostic component.

**Reply by rickghome** [reply]
Sure. But the point is about how to get the benefit of alpine for dynamically loaded components.

**Reply by ekwoka** [reply]
Here you don't express a dynamically loaded component. You express loading multiple of the same component with different initial data.

But the `clone the template` part doesn't seem relevant or helpful, btw. What does that accomplish?

**Comment by rickghome** [comment]
In step 3, a dynamic call is made from your webpage to the server using fetch (or other), to retrieve either an HTML partial, or json that you can templatize on the client.  In my use case, a user can add any number of these elements, and each will become an alpine component once they're added to the page.

**Reply by ekwoka** [reply]
Why not just, in that case, render the component on the server?

Or if it's just data, pass it as data to loops in alpine?

**Comment by rickghome** [comment]
The main point of this post, is that while the Alpine docs are excellent in general, they could be improved by offer some more best practices/recipes. In this case, how to load content dynamically from the web on your page, and properly initialize a component within that content.

**Reply by ekwoka** [reply]
In this case, what is happening and your solution are not best practices.

The best practice is load the component structure with alpine data for all such components (or inline the full data directly) and call it with arguments to instantiate that unique configured instance.

**Comment by rickghome** [comment]
Sorry ekwoka, that may work for pages where interactions don't change on the page, but we're talking about the case where a component can be added to th page dynamically by the user.  Best practice dictates that we design solutions to achieve what is necessary to satisfy the requirements. in this case, that means being able to add alpine components dynamically to a page. Where the component has data and behaviors.  

Have you ever built a solution where a component can be added dynamically to page?  How did you solve for this? Is there an example in the alpine docs?

**Reply by SimoTod** [reply]
I think ekwoka meant that you can simply return the component (HTML with your x-data already populated) and add it to the page without preloading the component template on the page but your solution is okay as well. 
It's not in the documentation because there are several ways to achieve that and none of them is necessary better than another. 
When you add additional opinionated snippets of code to the documentation, you also get people creating bugs and issues because the suggested "best practice" doesn't fit their use case, adding overhead for the maintainer.

**Reply by rickghome** [reply]
I see your point. It seems that the fact that I provided a scenario that included a way to inject your dynamic content is distracting from the real headline, which is how to achieve the use of alpine with dynamically loaded components.  I've rewritten my example to try to eliminate the confusion.

**Reply by ekwoka** [reply]
Okay, yes, this does make it more clear.
